Applications/Piano/KeysWidget.cpp

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2019-2020, William McPherson <willmcpherson2@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "KeysWidget.h"
+#include "AudioEngine.h"
+#include <LibGUI/GPainter.h>
+
+KeysWidget::KeysWidget(GWidget* parent, AudioEngine& audio_engine)
+    : GFrame(parent)
+    , m_audio_engine(audio_engine)
+{
+    set_frame_thickness(2);
+    set_frame_shadow(FrameShadow::Sunken);
+    set_frame_shape(FrameShape::Container);
+    set_fill_with_background_color(true);
+}
+
+KeysWidget::~KeysWidget()
+{
+}
+
+int KeysWidget::mouse_note() const
+{
+    if (m_mouse_down && m_mouse_note + m_audio_engine.octave_base() < note_count)
+        return m_mouse_note; // Can be -1.
+    else
+        return -1;
+}
+
+void KeysWidget::set_key(int key, Switch switch_key)
+{
+    if (key == -1 || key + m_audio_engine.octave_base() >= note_count)
+        return;
+
+    if (switch_key == On) {
+        ++m_key_on[key];
+    } else {
+        if (m_key_on[key] >= 1)
+            --m_key_on[key];
+    }
+    ASSERT(m_key_on[key] <= 2);
+
+    m_audio_engine.set_note_current_octave(key, switch_key);
+}
+
+static const int white_key_width = 24;
+static const int black_key_width = 16;
+static const int black_key_x_offset = black_key_width / 2;
+static const int black_key_height = 60;
```

Comment:
The right way to do consts like this in C++ is with `constexpr`

Reply:
Should I just slap `constexpr` on all of my `static const` variables? I do want them to be computed at compile time

Reply:
Basically, yes. And you can leave out the `static const` in that case (though ask @awesomekling if he prefers to leave them).

Reply:
You can leave out the `static const` like Sergey says yeah